en/lessons/specifics/benchmarking.md

```diff
@@ -0,0 +1,346 @@
+    * **none** - no unit scaling will occur. Durations will be displayed in microseconds, and ips counts will be displayed without units.
+
+## Inputs
+
+It's very important to benchmark your functions with data that reflects what that function might actually operate on in the real world. Frequently a function can have very different behavior on small sets of data versus large sets of data! This is where Benchee's `inputs` configuration option comes in. This allows you to test the same function but with as many different functions as you like, and then you can see the results of the benchmark with each of those functions.
+
+So, let's look at our original example again:
+
+```elixir
+list = Enum.to_list(1..10_000)
+map_fun = fn(i) -> [i, i * i] end
+
+Benchee.run(%{
+  "flat_map"    => fn -> Enum.flat_map(list, map_fun) end,
+  "map.flatten" => fn -> list |> Enum.map(map_fun) |> List.flatten end
+})
+```
+
+In that example we're only using a single list of the integers from 1 to 10,000. Let's update that to use a couple different inputs so we can see what happens with smaller and larger lists. So, open that file, and we're going to change it to look like this:
+
+```elixir
+map_fun = fn(i) -> [i, i * i] end
+inputs = %{
+  "small list" => Enum.to_list(1..100),
+  "medium list" => Enum.to_list(1..10_000),
+  "large list" => Enum.to_list(1..1_000_000)
+}
+
+Benchee.run(%{
+  "flat_map"    => fn(list) -> Enum.flat_map(list, map_fun) end,
+  "map.flatten" => fn(list) -> list |> Enum.map(map_fun) |> List.flatten end
+}, inputs: inputs)
+```
+
+You'll notice two differences. First, we now have an `inputs` map that contains the information for our inputs to our functions. We're passing that inputs map as a configuration option to `Benchee.run/2`.
+
+And since our functions need to take an argument now, we need to update our benchmark functions to accept an argument, so instead of:
+```elixir
+fn -> Enum.flat_map(list, map_fun) end
+```
+
+we now have:
+```elixir
+fn(list) -> Enum.flat_map(list, map_fun) end
+```
+
+Let's run this again using:
+
+```shell
+$ mix run benchmark.exs
+```
+
+Now you should see output in your console like this:
+
+```shell
+Operating System: macOS
+CPU Information: Intel(R) Core(TM) i5-4260U CPU @ 1.40GHz
+Number of Available Cores: 4
+Available memory: 8.589934592 GB
+Elixir 1.5.1
+Erlang 20.0
+Benchmark suite executing with the following configuration:
+warmup: 2.00 s
+time: 5.00 s
+parallel: 1
+inputs: large list, medium list, small list
+Estimated total run time: 2.10 min
+
+
+
+Benchmarking with input large list:
+Benchmarking flat_map...
+
+Benchmarking with input medium list:
+Benchmarking flat_map...
+
+Benchmarking with input small list:
+Benchmarking flat_map...
+
+Benchmarking with input large list:
+Benchmarking map.flatten...
+
+Benchmarking with input medium list:
+Benchmarking map.flatten...
+
+Benchmarking with input small list:
+Benchmarking map.flatten...
```

Comment:
You're running this from master, aren't you? ;P It's a bug I noticed the other day, it's supposed to look more like this:

```
Benchmarking with input Bigger:
Benchmarking flat_map...
Benchmarking map.flatten...

Benchmarking with input Medium:
Benchmarking flat_map...
Benchmarking map.flatten...

Benchmarking with input Small:
Benchmarking flat_map...
Benchmarking map.flatten...
```

But we might change that with the new structure as well...